### PR TITLE
utils: <math> formulas rendered to SVGs without using LaTeX tools

### LINKED ISSUE #1198: Generate SVG rather than GIF for embedded pictures
A successfull experiementation was done in https://github.com/BoboTiG/ebook-reader-dict/issues/1182#issuecomment-1027245425 about moving embedded pictures from GIF to SVG. Results are way better, so let's do the move.

We first need to ensure this works with PyGlossary and StarDict display.

Note: PyGlossary 4.4.2 or newer is required.

### LINKED ISSUE #1427: Use a custom docker image for tests
For each PR tests job, most of the time is taken by LateX installation. For instance, it takes about 2m40s to install it against 30s to run tests.

Maybe should we investigate the creation of a custom Docker image with LaTeX preinstalled. If so, I would be in favor of using a Debian-based **light** distribution, but I am open to any distribution as soon as tests are passing as-is (e.g: no modifications to be done on the source code).

### PR DESCRIPTION
**Pull request by BoboTiG**
Fixes #1427.
Fixes #1198.
Closes #1209.

Tests to pass before merging (the rendering is good, but not the display):

- [x] `$ python -m wikidict fr --gen-dict "cercle unité" --output issue-1427`
- [x] `$ python -m wikidict en --gen-dict "Wallis product,primitive recursion,Horner's rule" --output issue-1427`

### CODE REVIEW

**Comment by BoboTiG** [comment]
Question 1: Should we use a LRU cache on `render_formula()`?

Question 2: Should we store optimized SVGs for later runs? So we could run tests without calling the REST API.

**Comment by lasconic** [comment]
lru_cache: My understanding is that it will cache the results of a function when the function is called several times with the same args during a program run. It will not persists between run. I believe we don't call this function often with the same argument. Right ? So I would say lru_cache will not really speed up the process, but hey, we could try it.

store svgs: if it doesn't make the test moot, sure, we would store them.

File: wikidict/utils.py [hunk 1]
```diff
@@ -478,63 +499,64 @@ def process_templates(word: str, text: str, locale: str) -> str:
...
+
+
+def inline_svg(svg_raw: str) -> str:
+    """Return an optimized data-URI SVG string."""
+    # Alter units
+    # svg_raw = svg_raw.replace("ex", "em")
+
+    svg_optimized = scourString(svg_raw, options=SCOUR_OPTS)
+    return (
+        f'<img src="data:image/svg+xml;utf8,{quote(svg_optimized)}"/>'
```
[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py:537]: Not yet, good idea 👍

[the remark this answers is not in the document]
Comment (lasconic) [reply wikidict/utils.py:537]: The size is better but the display is messy :-/
screen_002

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py:537]: On the other hand, it displays perfectly here: 
screen_001

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py:537]: Maybe "cercle unité" contains raw text that scales badly. Will check the SVG content.

**Comment by BoboTiG** [comment]
> lru_cache: My understanding is that it will cache the results of a function when the function is called several times with the same args during a program run. It will not persists between run. I believe we don't call this function often with the same argument. Right ? So I would say lru_cache will not really speed up the process, but hey, we could try it.

I was thinking about duplicate formulas, but it will be near to zero at the end, mostly because the cache won't be shared accross locales. Let's forget that idea.

> store svgs: if it doesn't make the test moot, sure, we would store them.

It would make sense to store them: 2 HTTP calls saved for each formula. I'll implement it right now.

File: wikidict/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py]: Checked english words, the display is just perfect using `<svg>`. Let's hope we can fix the display issue for "cercle unité" (and potentially other words, but we mage a good progress still).

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py]: It seems `\\mathbb` is the problem. But I don't get why for now.

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py]: Hm interesting ... Saving the SVG to a file: it is displayed properly via the OS viewer.

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py]: OK, it's scoup that is doing something with IDs, and brwosers don't like.

[the remark this answers is not in the document]
Comment (BoboTiG) [reply wikidict/utils.py]: Fixed!

screen_001

screen_002

**Comment by BoboTiG** [comment]
@lasconic ready for final review.

Only one tiny concern about the potential size of `svg.py`, but I guess there are not thousands of formulas to keep in the cache.

**Comment by lasconic** [comment]
PNG master
screen_003
SVG this PR.
screen_004

It's crispy but it's too large no ? the last one is cut off.

**Comment by BoboTiG** [comment]
Let's see if other formulas are cut off too. I would say it's quite correct now (far better than what we used to have).

**Comment by BoboTiG** [comment]
Do you want to dig into scaling down the SVGs, maybe?

**Comment by BoboTiG** [comment]
I updated the cache with all formulas used in french. I'll update it when it gets merged.

**Comment by BoboTiG** [comment]
https://github.com/BoboTiG/ebook-reader-dict/pull/1432/commits/cb61adee2089b55c1941841c1569ab8c1f08c0a8 fixed the issue where we altered `<math>` formulas, especially on the SV dict.

**Comment by BoboTiG** [comment]
I'm finished with the cache.

Maybe before merging we could find a way to compress `svg_cache.py`? An idea:
- using `md5(formula)` for the keys
- compressing SVG data as `.tar.bz2` or something more powerfull from the standard modules. FTR `gzip` is badly perfoming.

**Comment by BoboTiG** [comment]
Ah, about the cut off, I didn't a big part of the picture was hidden. That's a shame.

**Comment by lasconic** [comment]
instead of md5, maybe SHA1 ? It seems wikimedia is doing this : https://github.com/wikimedia/restbase/blob/ecef17bda6f4efc0d6e187fb05b1eeb389bf7120/sys/post_data.js#L13

SVGZ is gzip no ? It doesn't perform well in term of speed ?

For the cut off, it's a pity but the real estate on kobo is so small... and we can't change the margins...

**Comment by BoboTiG** [comment]
> instead of md5, maybe SHA1 ? It seems wikimedia does this : https://github.com/wikimedia/restbase/blob/ecef17bda6f4efc0d6e187fb05b1eeb389bf7120/sys/post_data.js#L13

Weird, I can't replicate the hash.

```python
formula = "V^n"
print(sha1(formula.encode()).hexdigest())
```
It gives `1617fcdcebf4aa277a74b0c3bb4772ea65d8966f`, but `eb483cb4ed353a8b15938aa2d994cc5f0fb9055a` is returned by the service.

> SVGZ is gzip no ? It doesn't perform well in term of speed ?
> 
> For the cut off, it's a pity but the real estate on kobo is so small... and we can't change the margins... We could change the font size but it would need style for every word.

About SVGZ, I didn't try yet.

**Comment by BoboTiG** [comment]
No luck so far. I would just let the cache as-is until it becomes problematic.

**Comment by BoboTiG** [comment]
I finally could replicate the hash.

We need to use the normalized TeX, it's available in the 1<sup>st</sup> call:
```json
{
    "success": True,
    "checked": "V^{n}",
    "requiredPackages": [],
    "identifiers": ["V", "n"],
    "endsWithDot": False,
}
```
(See the `checked` key.)

Then, we need to use the full query, as you found out, but also without spaces between items:
```python
d = json.dumps({"q": formula, "type": "tex"}, indent=None)
d = d.replace(" ", "")
print(sha1(d.encode()).hexdigest())
```
I know, the POC is ugly, and will not work for all cases :)

**Comment by BoboTiG** [comment]
Well, we "just" need to fix the display issue, then we will be good to merge.
Do you have an idea on how to do that? Maybe using some CSS (sadly)?

**Comment by lasconic** [comment]
Which display issue ? If it's the cut out formula, then I don't think there is a fix... The screen of the kobo is just too small for some formula. If we make it smaller, it will become not readable.

**Comment by BoboTiG** [comment]
Formulas rendering is not a big feature of the dictionaries. If they're cut, I'm OK with that.

**Pull request merged**